Preserve whitespace in code blocks

silvenon opened the pull request:
I'm trying to write a test that reproduces #221. The idea is to mount the component exported by the fixture and use `textContent` to check if whitespace in the code block is preserved.

I'm having difficulties trying to load the module compiled by webpack as a React component and mounting it. Webpack gives me an object with a source text, but I'm not sure how to parse it as a Node module. If someone has an idea, let me know. 😅 

P. S. I'm not 100% certain that this is really an MDX issue, but the fact that I can't reproduce it without it suggests that it is.

silvenon commented:
I realized I could test this easily with runtime, so nvm, I think I got it. 😉

silvenon commented:
Ok, I managed to reproduce this in a test. 🎉

hugmanrique commented:
So would moving the prism plugin to compilers fix the issue? What's the main difference between mdPlugins and this new option?

silvenon commented:
`hastPlugins` operate on rehype tree, while `compilers` operate on JSX strings. Notice that between applying these two sets of plugins the tree is converted to JSX. `@mapbox/rehype-prism` can only work on rehype, so I think it has to stay in `hastPlugins`.

silvenon commented:
I (hopefully) figured it out! In the process of converting to JSX there was an exception with newlines introduced by #180. I added an exception to that exception (exceptionception) and converted newlines to JSX text nodes only for code inside `<pre>` tags.

@ifyoumakeit because you have a deeper understanding of that part of the code, could you take a look at this?

ifyoumakeit commented:
I can definitely can take a look!

ifyoumakeit commented on packages/mdx/mdx-hast-to-jsx.js:
```diff
@@ -47,7 +52,13 @@ function toJSX(node, parentNode = {}, options = {}) {
+    children = node.children.map(childNode => {
+      const childOptions = {
+        ...options,
+        preserveNewlines: preserveNewlines || node.tagName === 'pre',
```
👍 I think this makes sense, as long as there's information about this default behavior somewhere.

alexkrolick commented:
Nice, this was super annoying: https://github.com/alexkrolick/mdx-observable/pull/1

silvenon commented:
@ifyoumakeit thanks a lot! I’ll make sure to at least add comments.

@alexkrolick could you be more specific? What problem would these changes solve for you?

alexkrolick commented:
Actually I might be me misinterpreting the problem. I thought this would fix the fact that newlines inside render functions are interpreted as MD blocks instead of as JS code to evaluate. You can see how the formatting in my link requires adding comments or semicolons anywhere you want a blank line.

silvenon commented:
This PR unfortunately won't take care of that, it's for fenced code blocks only. We're painfully aware of the issue you mentioned, though.

silvenon commented:
Done! Polished it up a bit and added more comments to clarify what's going on.

@johno ready when you are. 😉

johno commented:
Thanks all, this is amazing!

ifyoumakeit commented on packages/mdx/test/index.test.js:
```diff
@@ -97,7 +103,7 @@ it('Should render elements without wrapping blank new lines', async () => {
   | Col1 | Col2  |`)
 
-  expect(result.includes('{`\n`}')).toBe(false)
+  expect(result).not.toContain('{`\n`}')
```
💅

silvenon replied:
![tenor](https://user-images.githubusercontent.com/471278/45098180-bea19680-b124-11e8-8559-255acbe68b05.gif)

devongovett commented:
Can we get this released?

johno commented:
Now published as `0.15.1`, apologies on the delay.
